### PR TITLE
Bluetooth: Samples: Connection time synchronization

### PR DESCRIPTION
**Pull request by rugeGerritsen**
This sample showcases microsecond accurate time-synchronization over Bluetooth connections.
It uses the controller anchor point information to create a common time reference between the central and peripheral clock.
    
A common module for Bluetooth clock will be extracted later.

### CODE REVIEW

File: samples/bluetooth/conn_time_sync/README.rst [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+.. note::
+   This sample requires less hardware resources when it is run on an nRF54L Series device compared to the nRF52 or nRF53 Series devices.
+   On an nRF54L Series device, only one GRTC channel and PPI channel is needed to set up accurate toggling of an LED.
+   On nRF52 and nRF53 Series devices, you also need one RTC peripheral, one TIMER peripheral, one EGU channel, three PPI channels, and one PPI group.
+
+
+Building and running
+********************
+
+.. |sample path| replace:: :file:`samples/bluetooth/iso_time_sync`
```
Comment (jostn) [inline samples/bluetooth/conn_time_sync/README.rst:100]: What is the intention with this? It shows up as nothing: https://github.com/rugeGerritsen/sdk-nrf/blob/conn-time-sync/samples/bluetooth/conn_time_sync/README.rst#building-and-running

File: samples/bluetooth/conn_time_sync/README.rst [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+  Received: c_anchor point (time=0, counter=0), c_trigger_time 72350720, value 0
+  Received: c_anchor point (time=0, counter=0), c_trigger_time 72486859, value 1
+  Received: c_anchor point (time=0, counter=0), c_trigger_time 72623303, value 0
+  Received: c_anchor point (time=0, counter=0), c_trigger_time 72759442, value 1
+  Received: c_anchor point (time=0, counter=0), c_trigger_time 72895733, value 0
+
+Dependencies
+*************
+
+This sample uses the following `sdk-nrfxlib`_ library:
```
Comment (jostn) [inline samples/bluetooth/conn_time_sync/README.rst:192]:
```suggestion
This sample uses the following `sdk-nrfxlib` library:
```

File: samples/bluetooth/conn_time_sync/README.rst [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+.. |sample path| replace:: :file:`samples/bluetooth/iso_time_sync`
+
+.. include:: /includes/build_and_run.txt
+
+Testing
+*******
+
+After programming the sample to both development kits, perform the following steps to test it:
+
+1. Connect to the kits with a terminal emulator (for example, `nRF Connect Serial Terminal`_).
```
Comment (jostn) [inline samples/bluetooth/conn_time_sync/README.rst:109]:
```suggestion
1. Connect to the kits with a terminal emulator (for example, `nRF Connect Serial Terminal`).
```

File: samples/bluetooth/conn_time_sync/README.rst (hunk 2 above)
Comment (rugeGerritsen) [reply samples/bluetooth/conn_time_sync/README.rst:192]: This is the way to use links with sphinx, see output on https://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-16550/nrf/samples/bluetooth/conn_time_sync/README.html

File: include/bluetooth/hci_vs_sdc.h [hunk 4]
```diff
@@ -392,6 +392,28 @@ int hci_vs_sdc_scan_accept_ext_adv_packets_set(
...
+ *
+ * For the complete API description, see sdc_hci_cmd_vs_set_event_start_task().
+ *
+ * @param[in]  params Input parameters.
+ *
+ * @return 0 on success or negative error value on failure.
+ */
+int hci_vs_sdc_set_event_start_task(const sdc_hci_cmd_vs_set_event_start_task_t *params);
+
+/** @brief Connection Anchor_Point Update Event Reports enable.
```
Comment (jostn) [inline include/bluetooth/hci_vs_sdc.h:405]: Anchor Point or Anchor-Point?

Comment (rugeGerritsen) [reply include/bluetooth/hci_vs_sdc.h:405]: Here I used what we use here: https://github.com/nrfconnect/sdk-nrfxlib/blob/main/softdevice_controller/include/sdc_hci_vs.h#L1726

File: samples/bluetooth/conn_time_sync/src/controller_time_nrf52.c [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+	IRQ_CONNECT(NRFX_IRQ_NUMBER_GET(NRF_RTC_INST_GET(2)), 0,
+		    (void *)NRFX_RTC_INST_HANDLER_GET(2), NULL, 0);
+#endif
+
+	nrfx_rtc_overflow_enable(&app_rtc_instance, true);
+	nrfx_rtc_tick_enable(&app_rtc_instance, false);
+	nrfx_rtc_enable(&app_rtc_instance);
+
+	/* To obtain the controller timestamp without modifying the state of RTC0,
+	 * we find the offset between RTC0 and our mirrored RTC.
```
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply samples/bluetooth/conn_time_sync/src/controller_time_nrf52.c:76]: Yes, we could do that.
I would vote for keeping it as is for two reasons:
1. It keeps the code more similar to what we have in the `iso_time_sync` sample:
2. The followup of this PR is to extract a common clock module which will use the `nrf_rtc_timer.h` API reusing a channel from RTC1. There we would need to do a similar trick.

File: samples/bluetooth/conn_time_sync/src/controller_time_nrf52.c [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+{
+	const uint32_t rtc_overflow_time_us = 512000000UL;
+
+	/* On the 52 series the RTC has to task to capture the current RTC value.
+	 * Therefore we cannot capture the TIMER and RTC value simultaneously.
+	 * Therefore we only use the RTC to read out the current time here.
+	 * This will result in an error of maximum one RTC tick.
+	 */
+
+	uint32_t rtc_ticks = nrf_rtc_counter_get(app_rtc_instance.p_reg);
```
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply samples/bluetooth/conn_time_sync/src/controller_time_nrf52.c:232]: Good catch!

File: samples/bluetooth/conn_time_sync/src/controller_time_nrf52.c [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+	while (sync_attempts > 0) {
+		sync_attempts--;
+
+		int32_t diff_measurement_1 = rtc_diff_get();
+
+		/* We need to wait half an RTC tick to ensure we are not measuring
+		 * the diff between the two RTCs at the point in time where their
+		 * values are transitioning.
+		 */
+		k_busy_wait(15);
```
[the remark this answers is not in the document]
Comment (rugeGerritsen) [reply samples/bluetooth/conn_time_sync/src/controller_time_nrf52.c:93]: I also thought that this was not necessary. I ran the following code a couple of times with and without the `k_busy_wait()`:
```c
int32_t diffs[300];

while (sync_attempts > 0) {
  sync_attempts--;
  diffs[sync_attempts] = rtc_diff_get();
  k_busy_wait(15);
}
print_diffs();
```
With the `k_busy_wait()`, there is sometimes a diff value that is off by 1, but never more than one in a row. Without the `k_busy_wait()`, there are sometimes a sequence of multiple values that are off.

File: samples/bluetooth/conn_time_sync/include/conn_time_sync.h [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+/** Obtain the current Bluetooth controller time.
+ *
+ * The timestamps are based upon this clock.
+ *
+ * @retval The current controller time.
+ */
+uint64_t controller_time_us_get(void);
+
+/** Sets the controller to trigger a PPI event at the given timestamp.
```
Comment (greg-fer) [inline samples/bluetooth/conn_time_sync/include/conn_time_sync.h:73]:
```suggestion
/** @brief Set the controller to trigger a PPI event at the given timestamp.
```

File: samples/bluetooth/conn_time_sync/include/conn_time_sync.h [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+ */
+uint32_t controller_time_trigger_event_addr_get(void);
+
+/** Initialize the module handling timed toggling of an LED.
+ *
+ * @retval 0 on success, failure otherwise.
+ */
+int timed_led_toggle_init(void);
+
+/** Toggle the led to the give value at the given timestamp.
```
Comment (greg-fer) [inline samples/bluetooth/conn_time_sync/include/conn_time_sync.h:91]:
```suggestion
/** @brief Toggle the LED using the given value at the given timestamp.
```

File: samples/bluetooth/conn_time_sync/include/conn_time_sync.h [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+ *
+ * @param value The LED value.
+ * @param timestamp_us The time when the led will be set.
```
Comment (greg-fer) [inline samples/bluetooth/conn_time_sync/include/conn_time_sync.h:94]: `us` = microseconds?
```suggestion
 * @param timestamp_us The time when the LED will be set, in microseconds.
```

File: samples/bluetooth/conn_time_sync/include/conn_time_sync.h [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+
+/** Get the address of the event that will trigger.
+ *
+ * @retval The address of the event that will trigger.
+ */
+uint32_t controller_time_trigger_event_addr_get(void);
+
+/** Initialize the module handling timed toggling of an LED.
+ *
+ * @retval 0 on success, failure otherwise.
```
Comment (greg-fer) [inline samples/bluetooth/conn_time_sync/include/conn_time_sync.h:87]:
```suggestion
 * @retval 0 Success.
 * @retval non-zero Failure.
```
As per the example in https://docs.nordicsemi.com/bundle/ncs-latest/page/nrf/dev_model_and_contributions/documentation/styleguide.html#typedefs

(global edit)

**Review by greg-fer — CHANGES_REQUESTED** [review]
First round of review focused on header files and the image.

File: samples/bluetooth/conn_time_sync/src/controller_time_nrf53_app.c [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+	 * It is possible to combine RTC and TIMER information here to get a more accurate
+	 * timestamp. That requires setting up a PPI channel to capture both values simultaneously.
+	 */
+
+	const uint64_t rtc_overflow_time_us = 512000000UL;
+
+	int key = irq_lock();
+
+	uint32_t rtc_ticks = nrf_rtc_counter_get(app_rtc_instance.p_reg);
+	uint32_t rtc_overflows = num_rtc_overflows;
```
Comment (jostn) [inline samples/bluetooth/conn_time_sync/src/controller_time_nrf53_app.c:213]: I believe we have the reverse problem now. The RTC may have overflowed after we disabled the interrupt on line 210, but this is not reflected in `num_rtc_overflows`

File: include/bluetooth/hci_vs_sdc.h [hunk 13]
```diff
@@ -392,6 +392,28 @@ int hci_vs_sdc_scan_accept_ext_adv_packets_set(
...
+ *
+ * For the complete API description,
+ * see sdc_hci_cmd_vs_conn_anchor_point_update_event_report_enable().
```
[the remark this answers is not in the document]
Comment (greg-fer) [reply include/bluetooth/hci_vs_sdc.h]: @gmarull , can you confirm?

File: samples/bluetooth/conn_time_sync/README.rst [hunk 14]
```diff
@@ -0,0 +1,222 @@
...
+* :file:`include/kernel.h`
+
+* :file:`include/sys/printk.h`
+* :file:`include/zephyr/types.h`
+* :ref:`zephyr:bluetooth_api`:
+
+* :file:`include/bluetooth/bluetooth.h`
+* :file:`include/bluetooth/conn.h`
+* :file:`include/bluetooth/hci.h`
+* :file:`include/bluetooth/scan.h`
```
Comment (greg-fer) [inline samples/bluetooth/conn_time_sync/README.rst:222]: Hopefully one last fix, to the list indentation:
```suggestion
* :file:`include/console.h`
* :ref:`zephyr:kernel_api`:

  * :file:`include/kernel.h`

* :file:`include/sys/printk.h`
* :file:`include/zephyr/types.h`
* :ref:`zephyr:bluetooth_api`:

  * :file:`include/bluetooth/bluetooth.h`
  * :file:`include/bluetooth/conn.h`
  * :file:`include/bluetooth/hci.h`
  * :file:`include/bluetooth/scan.h`
```

**Review by greg-fer — CHANGES_REQUESTED** [review]

**Review by jostn — APPROVED** [review]

**Pull request merged**